src/libpmemobj/ulog.c

```diff
@@ -588,6 +618,58 @@ ulog_inc_gen_num(struct ulog *ulog, const struct pmem_ops *p_ops)
 	VALGRIND_REMOVE_FROM_TX(&ulog->gen_num, gns);
 }
 
+/*
+ * ulog_free_by_ptr_next -- free all ulogs starting from the indicated one.
+ * Function returns 1 if any ulog have been freed or unpinned, 0 otherwise.
+ */
+int
+ulog_free_by_ptr_next(struct ulog *u, const struct pmem_ops *p_ops,
+		ulog_free_fn ulog_free)
+{
+	VEC(, uint64_t *) logs_past_first;
+	VEC_INIT(&logs_past_first);
+	int ret = 0;
+	struct ulog *ulog_head = u;
+
+	size_t next_offset;
+	while (u != NULL && ((next_offset = u->next) != 0)) {
+		if (VEC_PUSH_BACK(&logs_past_first, &u->next) != 0) {
+			/* this is fine, it will just use more pmem */
+			LOG(1, "unable to free transaction logs memory");
```

Comment:
You have to remove the user buffers, otherwise it will still be used if this fails.

You need to separate this function into two parts, in the first one you "unpin" all the user logs by changing the `next` to the user-buffer's next. After that, you simply free all the remaining buffers.